plat-rockchip: rk3588: add OTP_S support and HUK

edtubbs opened the pull request:
Add OTP_S support and HUK for Rockchip rk3588

jforissier commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
 	return TEE_SUCCESS;
 }
+
+static TEE_Result tee_otp_read_secure(uint32_t *pValue, uint32_t index,
```
s/pValue/value/, we don't use Hungarian notation

jforissier commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+	/* Setup read: index, count, command = READ */
+	uint32_t auto_ctrl_val = (index << ADDR_SHIFT) |
+				 (count << BURST_SHIFT) |
+				 CMD_READ;
```
`SHIFT_U32(index, ADDR_SHIFT)` etc.

jforissier commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+				   OTP_POLL_PERIOD_US,
+				   OTP_POLL_TIMEOUT_US)) {
+		return TEE_ERROR_BUSY;
+	}
```
{} not needed

jforissier commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+	return TEE_SUCCESS;
+}
+
+static TEE_Result tee_otp_write_secure(const uint32_t *pValue, uint32_t index,
```
s/pValue/value/

jforissier commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+		TEE_Result res = 0;
+
+		/* Setup write: curr_idx, command = WRITE */
+		uint32_t auto_ctrl_val = (curr_idx << ADDR_SHIFT) | CMD_WRITE;
```
SHIFT_U32

jforissier commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+	memset(buffer, 0, sizeof(buffer));
+	res = tee_otp_read_secure(buffer, HW_UNIQUE_KEY_INDEX,
+				  HW_UNIQUE_KEY_SIZE / sizeof(uint32_t));
+	if (res != TEE_SUCCESS) {
```
`if (res)`

jforissier commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+
+		/* Generate random 128-bit key from TRNG */
+		res = hw_get_random_bytes(rand_bytes, sizeof(rand_bytes));
+		if (res != TEE_SUCCESS) {
```
`if (res)`

etienne-lms commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+		panic("OTP_S base not mapped");
+
+	/* Check for invalid parameters or exceeding hardware burst limit */
+	if (!pValue || !count || count > BURST_SIZE)
```
Test also `index + count > MAX_INDEX` as for `tee_otp_write_secure()`?

etienne-lms commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+
+	/* Program OTP words */
+	for (uint32_t i = 0; i < count; i++) {
+		uint32_t old_val, new_val = 0;
```
1 variable per line.

etienne-lms commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+	for (uint32_t i = 0; i < count; i++) {
+		uint32_t old_val, new_val = 0;
+		uint32_t curr_idx = index + i;
+		TEE_Result res = 0;
```
prefer a known `TEE_Result` value instead of `0`, e.g. `TEE_SUCCESS` or `TEE_ERROR_GENERIC`.

etienne-lms commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+			return res;
+
+		/* Only program bits that are currently 0 (0->1) */
+		new_val = *pValue++ & ~old_val;
```
prefer 1 instruction per line (and let compiler optimize the sequence):
```c
		new_val = *pValue & ~old_val;
		pValue++;

etienne-lms commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
+{
+	TEE_Result res = 0;
+	uint32_t buffer[HW_UNIQUE_KEY_SIZE / sizeof(uint32_t)];
```
Prefer `= { }` here and remove the `memset()` instruction at line 338.

etienne-lms commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+	if (key_is_empty) {
+		/* HUK is empty => generate random key and program it */
+		uint8_t rand_bytes[HW_UNIQUE_KEY_SIZE];
+		uint32_t words[HW_UNIQUE_KEY_SIZE / sizeof(uint32_t)];
```
`= { }` for the 2 above lines

etienne-lms commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+		}
+
+		/* Convert random bytes -> 32-bit words */
+		memcpy(words, rand_bytes, sizeof(words));
```
Nitpicking: could have called `hw_get_random_bytes(buffer, sizeof(buffer))` (`buffer` is still a valid place where to generate the bytes).

etienne-lms commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* Program OTP words */
+	for (uint32_t i = 0; i < count; i++) {
```
Define `uint32_t i = 0;` at function entry instead of from this `for ()` instruction.
Ditto for line 397.

etienne-lms commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+	}
+
+	/* Copy HUK into hwkey->data */
+	memcpy(hwkey->data, buffer, HW_UNIQUE_KEY_SIZE);
```
Since the HUK is a sensible data, I suggest to wipe `buffer content` (`memzero_explicit()`).
By the way, maybe using straight `hwkey->data` to get/generate the HUK is enough, no strict need to use a temporary buffer.

etienne-lms reviewed (commented):
In commit comment, could you append parentheses to "tee_otp_get_hw_unique_key"?

DualTachyon commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+	/* Wait for RD_DONE bit using IO_READ32_POLL_TIMEOUT */
+	if (IO_READ32_POLL_TIMEOUT(base + OTP_S_INT_ST,
+				   int_status,
+				   int_status & OTP_S_RD_DONE_BIT,
```
You should also check the "ERROR" bit, otherwise you will be spending TIMEOUT time waiting for the call to return. There are some OTP addresses that will return such an error.

DualTachyon commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+			return TEE_ERROR_BUSY;
+		}
+
+		/* Clear INT status bits */
```
You could clear right after the read, right now it only clears on success.

edtubbs commented:
Thanks everyone! I've addressed all the feedback.

@etienne-lms I'm still using `buffer` instead of casting `hwkey->data` to `uint32_t *` to avoid alignment warnings.

edtubbs commented:
Thanks!

edtubbs commented:
All the feedback has been addressed. Let me know if there's anything else. Thanks!

edtubbs commented:
@etienne-lms @jforissier I've addressed your feedback.

etienne-lms commented on core/arch/arm/plat-rockchip/platform_rk3588.c:
```diff
@@ -160,3 +192,208 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+		new_val = *value & ~old_val;
+		value++;
+		if (!new_val)
+			continue;
```
I meant if target word to write expects bit changes from 1 to 0 (seems not possible), should not this driver report a `TEE_ERROR_GENERIC` or like? The implementation makes them to be ignored.

etienne-lms replied:
Sorry, indeed this is addressed from line 292.
This comment thread is resolved.

edtubbs commented:
@etienne-lms This is already handled: if `old_val` has `1`s where `*value` has `0`s, it returns `TEE_ERROR_GENERIC`. https://github.com/OP-TEE/optee_os/blob/c7a413b1dd2cb1c99e4bf8a664044aae5f526cd6/core/arch/arm/plat-rockchip/platform_rk3588.c#L291-L295

etienne-lms reviewed (approved):
All good to me.

edtubbs commented:
@jforissier do you approve of the updates? Let me know if anything else is needed—thanks!

jforissier commented:
@edtubbs yes, this looks ready for merging. Thanks and sorry for the delay.
